cpp/src/quantiles/groupby.cu

```diff
@@ -0,0 +1,264 @@
+  decltype(idx_data.begin()) result_end;
+
+  if (nullable) {
+    auto comp = row_equality_comparator<true>(*device_input_table, true);
+    result_end = thrust::unique_copy(
+      thrust::device, counting_iter, counting_iter + _num_keys,
+      idx_data.begin(), transform_row_eq_comparator<true>{comp, sorted_order});
+  } else {
+    auto comp = row_equality_comparator<false>(*device_input_table, true);
+    result_end = thrust::unique_copy(
+      thrust::device, counting_iter, counting_iter + _num_keys,
+      idx_data.begin(), transform_row_eq_comparator<false>{comp, sorted_order});
+  }
+
+  gdf_size_type num_groups = thrust::distance(idx_data.begin(), result_end);
+  _group_ids = std::make_unique<index_vector>(idx_data.begin(), idx_data.begin() + num_groups);
+
+  return *_group_ids;
+}
+
+rmm::device_vector<gdf_size_type> const& groupby::group_labels() {
+  if (_group_labels)
+    return *_group_labels;
+
+  // Get group labels for future use in segmented sorting
+  _group_labels = std::make_unique<index_vector>(_num_keys);
+
+  auto& group_labels = *_group_labels;
+  thrust::fill(group_labels.begin(), group_labels.end(), 0);
+  auto group_labels_ptr = group_labels.data().get();
+  auto group_ids_ptr = group_indices().data().get();
+  thrust::for_each_n(thrust::make_counting_iterator(1),
+                    group_indices().size() - 1,
+                    permutation_label_setter{group_labels_ptr, group_ids_ptr});
+  thrust::inclusive_scan(thrust::device,
+                        group_labels.begin(),
+                        group_labels.end(),
+                        group_labels.begin());
+
+  return group_labels;
+}
+
+gdf_column const& groupby::unsorted_labels() {
+  if (_unsorted_labels)
+    return *_unsorted_labels;
+
+  _unsorted_labels = std::make_unique<gdf_column>(
+    allocate_column(gdf_dtype_of<gdf_size_type>(),
+                    key_sort_order().size));
+  auto& unsorted_labels = *_unsorted_labels;
+  cudaMemset(unsorted_labels.valid, 0,
+              gdf_num_bitmask_elements(unsorted_labels.size));
+  
+  gdf_column group_labels_col{};
+  gdf_column_view(&group_labels_col, 
+                  const_cast<gdf_size_type*>(group_labels().data().get()), 
+                  nullptr,
+                  group_labels().size(), 
+                  gdf_dtype_of<gdf_size_type>());
+  cudf::table t_sorted_labels{&group_labels_col};
+  cudf::table t_unsorted_labels{&unsorted_labels};
+  cudf::detail::scatter(&t_sorted_labels,
+                        reinterpret_cast<gdf_size_type*>(key_sort_order().data),
+                        &t_unsorted_labels);
+  return unsorted_labels;
+}
+
+
+std::pair<gdf_column, rmm::device_vector<gdf_size_type> >
+groupby::sort_values(gdf_column const& val_col) {
+  auto idx_col = allocate_column(gdf_dtype_of<gdf_index_type>(),
+                                _key_table.num_rows(),
+                                false);
+
+  // We need a table constructor that can take const initializer list
```

Comment:
Agree. I have run into this multiple times.

Reply:
Actually, logically, this shouldn't be a problem when we're constructing a `const table`. I don't think there's a way to do that though https://stackoverflow.com/questions/49151393/constructor-for-const-object

Reply:
Now that we've all learned our lesson, please delete the comment, @devavret  :)